### ISSUE TITLE
[Security] AutoSpill (AutoFill Vunerability in WebView)

### ISSUE THREAD

**Issue by Vinfall** [Kunzisoft/KeePassDX#1716]
**Describe the bug**

Black Hat Europe 2023 presentation ([PDF version](https://i.blackhat.com/EU-23/Presentations/EU-23-Gangwal-AutoSpill-Zero-Effort-Credential-Stealing.pdf)) introduced a AutoFill service vulnerability named _AutoSpill_ to circumtivate autofill mechanism.

It is ultimately caused by WebView's behavior and seems quite worrying with the combination of JavaScript injection (see results below).

**To Reproduce**

Steps to reproduce the behavior: See PDF p43-50.

**Expected behavior**

When in WebView, AutoFill should only get triggered in desired URL, and invulnerable to JavaScript injection.

**Android:**

| Model                      | Type       | Android version | Android security patch |
| -------------------------- | ---------- | --------------- | ---------------------- |
| Poco F1                    | Smartphone | Android 10      | December 2020          |
| Samsung Galaxy Tab S6 Lite | Tablet     | Android 11      | January 2022           |
| Samsung Galaxy A52         | Smartphone | Android 12      | April 2022             |

**Additional context**

Personally I don't use AutoFill in WebView so am not affected but I'd like to know how much KeePassDX is affected and suggest a fix. Maybe you can temporarily disable AutoFill in WebView (like 1Password does) while the more sophisticated fix is on the way. You can find more details regarding the vulnerability in the PDF (notably p28-57) and reaction of other password manager on [TechCrunch](https://techcrunch.com/2023/12/06/your-mobile-password-manager-might-be-exposing-your-credentials/).
 - Browser for Autofill: Android System WebView (no idea about version, but I think it's tested on latest version as of Dec 2022)

**Comment by notahomosapien** [comment]
I think the fix for now is as simple as adding `com.google.android.webview` by default in the Application Blocklist in Autofill settings. If the user tries to remove `com.google.android.webview` from the Application Blocklist, an warning popup should be shown about this vulnurebility.

**Comment by Vinfall** [comment]
I think so, maybe adding (legacy) [bromite](https://github.com/bromite/bromite/releases) package names (i.e. `org.bromite.webview` & `com.android.webview`) since they offered (quite outdated) System WebView as well.

**Comment by J-Jamet** [comment]
You're right, it's a subject that needs to be taken seriously. 
First of all, as you said, you need to add WebViews to the blocklist.
But I'd like to understand the problem in more detail with a real case study to find out if it's possible to filter correctly with AutofillId and WebDomain. Do you have the link to the demonstration video?

**Comment by Vinfall** [comment]
> Do you have the link to the demonstration video?

Sadly no, all I have is the slides above, [briefing info](https://www.blackhat.com/eu-23/briefings/schedule/#autospill-zero-effort-credential-stealing-from-mobile-password-managers-34420) and numerous news report. It seems Black Hat Europe 2023 videos are not uploaded to [their YouTube playlist](https://www.youtube.com/@BlackHatOfficialYT/playlists) yet, and I can't find any PoC about this too.

However, from the author's GitHub I find the related paper on CODASPY '23 titled _[AutoSpill: Credential Leakage from Mobile Password Managers](https://dl.acm.org/doi/10.1145/3577923.3583658)_, I hope this would help.

**Comment by mooglestiltzkin** [comment]
any suggestions what settings to set in keepass to avoid this issue? or we can't do anything until keepassdx releases an update for this.

**Comment by Vinfall** [comment]
The most obvious workaround is to avoid using AutoFill in WebView, you can continue autofilling other fields (_if you can tell them apart_, that is, as some apps are basically a wrapper of WebView).

The linked PDF has some examples you should avoid:
![webview-0](https://github.com/Kunzisoft/KeePassDX/assets/91039000/6b9bb604-19e2-4251-a21b-5c9aa6b0df76)
![webview-1](https://github.com/Kunzisoft/KeePassDX/assets/91039000/aeb4297f-1d8b-41c6-8b55-149cda5ebe75)

**Comment by notahomosapien** [comment]
> I think so, maybe adding (legacy) [bromite](https://github.com/bromite/bromite/releases) package names (i.e. `org.bromite.webview` & `com.android.webview`) since they offered (quite outdated) System WebView as well.

So just dynamically adding whatever Webview is available to the blocklist seems to be the best solution

**Comment by shuvashish76** [comment]
For reference:
https://infosec.exchange/@AlesandroOrtiz/111574173202352335
https://www.bleepingcomputer.com/news/security/autospill-attack-steals-credentials-from-android-password-managers/

**Comment by J-Jamet** [comment]
Since version 4.3.0, webviews are simply blocked.

**Comment by J-Jamet** [comment]
https://github.com/heylogin/android_autospill_tester
The blocking code of webviews is revert because it causes too many recognition issues.
I tested it with the [android_autospill_tester](https://github.com/heylogin/android_autospill_tester) app, and each domain is handled separately, so normally there should be no issue if you pay attention to the domain that is registered.
But if you want to run more tests and let me know if you see any other solutions that are more specific to the problem, please don't hesitate.

**Comment by Vinfall** [comment]
> normally there should be no issue if you pay attention to the domain that is registered.

Agreed. Honestly speaking, if a malicious app makes its way to my phone, it would have many other more effective ways to collect my data. This is not a critical issue and I have not seen it exploited in the wild over the past two years.

Ironically, for some apps not using WebView but their own implementation, things are usually worse but that's definitely out the scope of password manager.
